core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1767,7 +1767,7 @@ class Log(@volatile var dir: File,
     if (config.retentionSize < 0 || size < config.retentionSize) return 0
     var diff = size - config.retentionSize
     def shouldDelete(segment: LogSegment, nextSegmentOpt: Option[LogSegment]) = {
-      if (diff - segment.size >= 0) {
+      if (diff >= 0) {
```

Comment:
I agree this behavior is a little surprising. I'm guessing the previous logic was trying to be conservative about segment deletion. It treats the retention size more as a lower bound on the retained size rather than an upper bound as the name suggests. Changes in behavior like this can be a little subtle. For example, the old logic effectively ensures that the active segment does not get deleted due to retention size. So even if you set retention size to 0, then we still retain one segment's worth of data. Is this behavior worth preserving? I'm not sure. An alternative is that we document the existing behavior more clearly. It shouldn't be the case that we have to guess whether this logic is intentional or not. 

cc @junrao Maybe you have some insight about the existing logic?

Reply:
@Evelyn-Bayes : Thanks for the PR. I looked at the original jira (https://issues.apache.org/jira/browse/KAFKA-70), when this feature was original introduced. There were some discussions on why we implemented the code this way. Basically, the concern was that if we do what this PR does, then we may be deleting the data too soon for the consumer to get a chance to read it. So, perhaps we could adjust the doc accordingly?

Reply:
@junrao I think you're right and I understand erring on the side of caution.
Although technically accurate I could see how this could create more problems than it solves.

I'll follow this up with an update to the documentation.

Reply:
@hachikuji might be easier to get you to take a look :)

Reply:
KAFKA-70, wow.